### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by danielaRiesgo**
## Sumary ##
**Trello card**: https://trello.com/c/gFXtp9T0/32-documentar

Documenting all framework elements.
And some small refactoring that came up:
    Changed DefaultXXXViewConfiguration, to a XXXViewConfiguration struct with default init parameters.
    Added/Removed view optional elements that were missing/unnecessary.

## Test Plan ##

  * Test all the app to check nothing changed.
  * Both login and signup succeed if using (email: "example@mail.com", password: "password")

## Known issues ##
Some session service issues are left as TODO because decisions are to be made (in another PR) about it.

### CODE REVIEW

**Comment by woloxci** [comment]
:+1: There are no linter violations.

File: Authentication/AuthenticationBootstrapper.swift [hunk 1]
```diff
@@ -63,10 +64,10 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
         }
     }
 
-    /**
+    /*
         Bootstraps your project with the authentication framework,
```
Comment (guidomb) [inline Authentication/AuthenticationBootstrapper.swift]: What about something like this?

```
Bootstraps your application with the authentication framework,
starting with the login screen if no user is logged in.
Otherwise, it runs your application's Main View Controller.
```

File: Authentication/AuthenticationBootstrapper.swift [hunk 2]
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
 
 extension AuthenticationBootstrapper: SignupControllerTransitionDelegate {
     
+    /*
+         Function that reacts to the user pressing "Log In"
+         in the signup screen.
+         It will pop the signup controller from the navigation controller,
+         to return to login screen.
+     */
     public func onLogin(controller: SignupController) {
```
Comment (guidomb) [inline Authentication/AuthenticationBootstrapper.swift]: this method should be final

File: Authentication/AuthenticationBootstrapper.swift [hunk 3]
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
+        // The authentication framework starts the process with a navigation controller.
+        controller.navigationController!.pushViewController(signupController, animated: true)
     }
     
+    /*
+         Function that reacts to the user pressing "Recover Password"
+         in the login screen.
+         It will push the new controller in the navigation controller.
+     */
     public func onRecoverPassword(controller: LoginController) {
```
Comment (guidomb) [inline Authentication/AuthenticationBootstrapper.swift]: this method should be final

File: Authentication/AuthenticationBootstrapper.swift [hunk 4]
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
+
+
 extension AuthenticationBootstrapper: LoginControllerTransitionDelegate {
     
+    /*
+         Function that reacts to the user pressing "Sign Up" in the
+         login screen.
+         It will push the new controller in the navigation controller.
+     */
     public func onSignup(controller: LoginController) {
```
Comment (guidomb) [inline Authentication/AuthenticationBootstrapper.swift]: this method should be final

File: Authentication/Login/Model/User.swift [hunk 5]
```diff
@@ -8,7 +8,11 @@
 
 import Foundation
 
-
+/*
+     Represents all properties needed from
+     a user for it to be compatible with
+     the authentication framework.
+ */
 public protocol UserType {
```
Comment (guidomb) [inline Authentication/Login/Model/User.swift]: Maybe for another PR but I think we can remove this protocol.

File: Authentication/Login/View/LoginController.swift [hunk 6]
```diff
@@ -12,13 +12,19 @@ import Rex
...
-    view model and informing the log in controller delegate when certain
+    view model and informing the login controller delegate when certain
     events occur for it to act upon them.
-    If there are more than one errors in a field, the controller presents
-    only the first one in the errors label.
+    If there are more than one validation error in a field, the controller
+    presents only the first one in the errors label.
+ 
+    If wanting to use the default LoginController with some customization,
+    you should not override the `createLoginController` method of the Bootstrapper,
```
Comment (guidomb) [inline Authentication/Login/View/LoginController.swift]: Users cannot override `createLoginController` because its visibility is `internal`.

Comment (guidomb) [inline Authentication/Login/View/LoginController.swift]: Also I don't think this should be explained here because all those methods are part of the bootstrapper. I think that the documentation should only "talk" about the class being documented not how other classes use it.

File: Authentication/Login/View/LoginViewConfiguration.swift [hunk 7]
```diff
@@ -6,8 +6,12 @@
 //  Copyright © 2016 Wolox. All rights reserved.
 //
 
-import Foundation
-
+/*
+    Represents the configurations able to
```
Comment (guidomb) [inline Authentication/Login/View/LoginViewConfiguration.swift]: What about `Represents the configurable parameters of a view that conforms to LoginViewType`?

**Comment by guidomb** [comment]
I've made a quick pass and left some comments. Haven't got the time to do deep review. I guess that the documentation will improve over time with the usage of the library so I think it is a very good starting point. We should see if other devs can properly configure the lib without much guidance and only using the docs to be sure that they are good.

File: Authentication/AuthenticationBootstrapper.swift (hunk 1 above)
Comment (danielaRiesgo) [inline Authentication/AuthenticationBootstrapper.swift]: The only thing  is that then it will start with login or signup according to what the user decides. `authentication process`? Or maybe explaining exactly that: `starting with the login or signup screen (according to you preferences) if no user is logged in.`

File: Authentication/Login/View/LoginController.swift (hunk 6 above)
Comment (danielaRiesgo) [inline Authentication/Login/View/LoginController.swift]: I agree. But as a framework I think this kind of explanations are nice to have, although maybe this is not the place. Where could we move it?

Comment (guidomb) [inline Authentication/Login/View/LoginController.swift]: Yeah that is true. My main concern is not being "coupled" with other parts of the library in a method or class documentation because, most probably, if something on how this class is used changes (not necessarily this class' API) then this docs would get out of date. Maybe the best place for high level usage documentation is the README, a tutorial or the Wiki.

**Comment by woloxci** [comment]
:+1: There are no linter violations.

File: Authentication/SignUp/View/SignupController.swift [hunk 8]
```diff
@@ -8,6 +8,21 @@
...
+
+/*
+     Signup View Controller that takes care of managing the signup, from
+     validating all fields, to binding a signup view to the view model
+     and informing the signup controller delegate when certain events 
+     occur for it to act upon them.
+     If there are more than one validation error in a field, the controller
+     presents only the first one in the errors label.
+     
+     If wanting to use the default SignupController, you should not override
```
Comment (woloxci) [inline Authentication/SignUp/View/SignupController.swift]: WARNING - TODOs and FIXMEs should be avoided.

**Comment by woloxci** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

File: Authentication/AuthenticationBootstrapper.swift [hunk 9]
```diff
@@ -8,7 +8,7 @@
 
 import Foundation
 
-/**
+/*
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: Why did you change this? If you remove the second *, the user of this class will not be able to see the description.

Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: with `/**`:
screen shot 2016-07-27 at 21 51 24

Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: with `/*`( just one *):
screen shot 2016-07-27 at 21 52 36

File: Authentication/AuthenticationBootstrapper.swift [hunk 10]
```diff
@@ -31,7 +34,7 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
         return sessionService.currentUser.value
     }
 
-    /**
+    /*
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: Why did you change this? If you remove the second *, the user of this class will not be able to see the description.

File: Authentication/AuthenticationBootstrapper.swift [hunk 11]
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
             _window.rootViewController = UINavigationController(rootViewController: createLoginController())
         }
     }
+}
 
-    /**
+// MARK: - Login Functions
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: You could leave mark, but extensions already make it clear

File: Authentication/AuthenticationBootstrapper.swift [hunk 12]
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
+public extension AuthenticationBootstrapper {
+    
+    /*
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

File: Authentication/AuthenticationBootstrapper.swift [hunk 13]
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
+         the authentication process.
+         
+         - Returns: A valid login controller to use.
+     */
+    internal func createLoginController() -> LoginController {
+        let configuration = createLoginControllerConfiguration()
+        return LoginController(configuration: configuration)
+    }
+    
+    /*
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

File: Authentication/AuthenticationBootstrapper.swift [hunk 14]
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
...
         - Attention: Override this method for customizing the criteria of email and password validity.
+        You can use email as username if you use the correct validator for it.
     */
     public func createLogInCredentialsValidator() -> LoginCredentialsValidator {
         return LoginCredentialsValidator()
     }
 
-    /**
-         Creates the LogInViewModel to use in the authentication process logic,
+    /*
```
Comment (FranDepascuali) [inline Authentication/AuthenticationBootstrapper.swift]: Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

**Comment by FranDepascuali** [comment]
@danielaRiesgo there are a lot of places in which you changed /** to /*. This are different things: with two **, it is documented (meaning that everyone who uses that function/class will be able to see the description). With just one *, it is just a comment, so you will only be able to see it if you are reading the implementation. I think you want the two ** (`/**`)

File: Authentication/ViewProtocols.swift [hunk 15]
```diff
@@ -6,6 +6,10 @@
 //  Copyright © 2016 Wolox. All rights reserved.
 //
 
+/*
+     Represents something that can give us an
+     UIView to show and render.
+ */
```
Comment (FranDepascuali) [inline Authentication/ViewProtocols.swift]: I would write just: 
```swift
Represents a component that can be visually displayed on the screen
```

**Comment by FranDepascuali** [comment]
@danielaRiesgo 
The only small detail is that there are some places in which you describe, for example a protocol, with something that it exposes or some detail about its implementation.
I think that those particular details should go on the things that it exposes and not on the protocol itself. The comment for a protocol should just describe what it does (or means). 

Having that said, it is just a small detail (and I don't even know if I'm right), but I like very much how you document functions, protocols and classes at all. I think that you are one of the best of us in that sense, very good 👍

**Comment by danielaRiesgo** [comment]
I changed to `/*` because I saw it maid valid_docs violations disappear. Now I understand why. I changed it back, and tried to solve all violations, but there are two I can't figure out what the problem is.

**Comment by danielaRiesgo** [comment]
@FranDepascuali there I checked the protocols' documentation, and tried to improve it, but I don't know if I got what you meant.

**Comment by woloxci** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

File: AuthenticationDemo/en.lproj/Main.strings [hunk 16]
```diff
@@ -1,3 +1,3 @@
 
-/* Class = "UILabel"; text = "Esta es la pantalla de tu aplicación !"; ObjectID = "VF1-gk-0yi"; */
+/** Class = "UILabel"; text = "Esta es la pantalla de tu aplicación !"; ObjectID = "VF1-gk-0yi"; */
```
Comment (FranDepascuali) [inline AuthenticationDemo/en.lproj/Main.strings]: I think you want this to be a comment

**Comment by FranDepascuali** [comment]
correct that detail and 👍

**Comment by woloxci** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

**Comment by guidomb** [comment]
@danielaRiesgo Did you rebase this PR? If so try not to do that once a review is in progress. If think I've already approved this one so if you just fixed the comments made by @FranDepascuali is ready to be merged. I am not reviewing this again so I trust that is the case. 👍
